app/javascript/components/routers-form/index.jsx

```diff
@@ -0,0 +1,163 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import MiqFormRenderer from '@@ddf';
+import { Loading } from 'carbon-components-react';
+import createSchema from './RoutersForm.schema';
+import miqRedirectBack from '../../helpers/miq-redirect-back';
+import { API } from '../../http_api';
+
+const RoutersForm = ({ routerId }) => {
+  const [{
+    initialValues, isLoading, fields, subnets,
+  }, setState] = useState({
+    isLoading: !!routerId, fields: [], subnetSelect: [], subnets: [],
+  });
+  const submitLabel = !!routerId ? __('Save') : __('Add');
+
+  let network = '';
+
+  const callAPI = (value) => {
+    API.get(`/api/cloud_subnets?expand=resources&attributes=name,ems_ref&filter[]=cloud_network_id=${value}`).then((data) => {
+      let subnets = data.resources;
+      subnets = subnets.map(({ id, name }) => ({ label: name, value: id }));
+      setState((state) => ({
+        ...state,
+        subnets,
+      }));
+    });
+  };
+
+  const loadSchema = (appendState = {}) => ({ data: { form_schema: { fields } } }) => {
+    Object.assign(fields[2].fields[2], {
+      resolveProps: (props, { meta, input }) => {
+        if (input.value && meta.active) {
+          if (input.value !== network) {
+            callAPI(input.value);
+            network = input.value;
+          }
+        }
+      },
+    });
+    setState((state) => ({
+      ...state,
+      ...appendState,
+      fields,
+    }));
+  };
+
+  const emptySchema = (appendState = {}) => {
+    const fields = [];
+    const subnetSelect = [];
+    setState((state) => ({
+      ...state,
+      ...appendState,
+      fields,
+      subnetSelect,
+    }));
+  };
+
+  useEffect(() => {
+    if (routerId && isLoading) {
+      miqSparkleOn();
+      API.get(`/api/network_routers/${routerId}?attributes=cloud_network.cloud_subnets.ids,external_gateway_info`)
+        .then((initialValues) => {
+          initialValues.enable = !!initialValues.extra_attributes.external_gateway_info;
```

Comment:
@GilbertCherrie Is it possible to move some of the logic to a small functions and then move to helpers file?